### PR TITLE
coordinate bulk operation retries once for every cluster node on each no...

### PR DESCRIPTION
**Pull request by mfelsche**
...de to optimize waiting times for retries

### CODE REVIEW

File: sql/src/main/java/org/elasticsearch/action/bulk/BulkRetryCoordinator.java [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        trace("execute shard request %d", updateRequest.shardId());
+        transportActionProvider.transportShardUpsertActionDelegate().execute(updateRequest, new BulkActionListener<>(listener, updateRequest));
+    }
+
+    public void retry(final ShardUpsertRequest request,
+                      boolean repeatingRetry,
+                      final RequestActionListener<ShardUpsertRequest, ShardUpsertResponse> retryListener) {
+        trace("doRetry");
+        final TransportShardUpsertActionDelegate delegate = transportActionProvider.transportShardUpsertActionDelegate();
+        final RetryBulkActionListener<ShardUpsertRequest, ShardUpsertResponse> retrybulkActionListener = new RetryBulkActionListener<>(retryListener, request);
```
Comment (seut) [inline sql/src/main/java/org/elasticsearch/action/bulk/BulkRetryCoordinator.java]: minor: change to retryBulkActionListener

File: sql/src/main/java/org/elasticsearch/action/bulk/BulkRetryCoordinatorPool.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        super(settings);
+        this.transportActionProvider = transportActionProvider;
+        this.coordinatorsByShardId = new HashMap<>();
+        this.coordinatorsByNodeId = new HashMap<>();
+        this.clusterService = clusterService;
+    }
+
+    @Nullable
+    public BulkRetryCoordinator coordinator(ShardId shardId) {
+        BulkRetryCoordinator coordinator = coordinatorsByShardId.get(shardId);
```
Comment (seut) [inline sql/src/main/java/org/elasticsearch/action/bulk/BulkRetryCoordinatorPool.java]: this needs to be syncronized/thread-safe

File: sql/src/main/java/org/elasticsearch/action/bulk/BulkShardProcessor.java [hunk 3]
```diff
@@ -146,32 +148,35 @@ public boolean add(String indexName,
...
     }
 
-    private void partitionRequestByShard(String indexName,
-                                         String id,
-                                         Row row,
-                                         @Nullable String routing,
-                                         @Nullable Long version) {
-        ShardId shardId = clusterService.operationRouting().indexShards(
+    private ShardId shardId(String indexName, String id, @Nullable String routing) {
+        return clusterService.operationRouting().indexShards(
```
Comment (seut) [inline sql/src/main/java/org/elasticsearch/action/bulk/BulkShardProcessor.java]: why not just return new ShardId(index, id) ?

File: sql/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java [hunk 4]
```diff
@@ -271,7 +271,7 @@ private void validateInsertPartitionedTable() {
     public void testMultiValueInsertPartitionedTable() throws Exception {
         execute("create table parted (id integer, name string, date timestamp)" +
                 "partitioned by (date)");
-        ensureYellow();
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java]: ensureYellow should be enough, revert

**Comment by seut** [comment]
travis fails through invalid routing, pls check

**Comment by seut** [comment]
ack if travis succeed ;)

**Comment by mfelsche** [comment]
For the record:

screen shot 2015-04-17 at 3 29 37 pm

**Pull request merged**